frontend/src/scenes/insights/InsightPageHeader.tsx

```diff
@@ -330,7 +332,19 @@ export function InsightPageHeader({ insightLogicProps }: { insightLogicProps: In
                             <LemonButton
                                 tooltip={showQueryEditor ? 'Hide JSON editor' : 'Edit as JSON'}
                                 type={'secondary'}
-                                onClick={toggleQueryEditorPanel}
+                                onClick={() => {
+                                    // for an existing insight in view mode
+                                    if (hasDashboardItemId && insightMode !== ItemMode.Edit) {
+                                        // enter edit mode
+                                        setInsightMode(ItemMode.Edit, null)
+
+                                        // exit early if query editor doesn't need to be toggled
+                                        if (showQueryEditor === false) {
+                                            return
+                                        }
+                                    }
```

Comment:
We should never show a button that's a no-op, it should either have a `disabledReason`, or ideally _just work_

Reply:
Just fixed a small issue with the toggling behaviour and now the button should never be a no-op:

```
for saved insights (hasDashboardItemId):
    when both off -> both on
    when edit off -> edit on
    when query off -> query on
    when both on -> query off
for unsaved insights
    query off -> query on
    query on -> query off
```

https://user-images.githubusercontent.com/1851359/235087232-a7a2e0f1-bb10-44c7-92fc-562a0f5e3041.mov